ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,262 @@
+		ret = verify_cert_extensions(store->certs->buffer,
+					     store->certs->length);
+		dlclose(handle);
+	}
+
+	fprintf(stderr, "Verifying certificate extensions ... %s\n",
+		ret == 0 ? "Success" : "Failure");
+	return !ret;
+}
+
+void *enclave_tls_init(struct enclave_tls_ctx *ctx)
+{
+	ctab.generate_certificate = &generate_x509;
+
+	if (ctx->quote_type == EPID) {
+		ctx->handle = dlopen(LIBENCLAVE_TLS_SGX, RTLD_LAZY);
+		if (NULL == ctx->handle) {
+			printf("open error:%s\n", dlerror());
+			return -1;
+		}
+#define DLSYM(fn)									\
+		do {									\
+			ftab.fn = dlsym(ctx->handle, #fn "_sgx");				\
+		} while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	} else {
+#define DLSYM(fn)                                                                       \
+                do {                                                                    \
+                        ftab.fn = &fn;                                                  \
+                } while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	}
+
+	ftab.wolfSSL_Debugging_ON();
+
+	ftab.wolfSSL_Init(ctx->quote_type);
+
+	WOLFSSL_CTX *wolf_ctx;
+	if (ctx->is_server)
+		wolf_ctx = ftab.wolfSSL_CTX_new(ftab.wolfTLSv1_2_server_method());
+	else
+		wolf_ctx = ftab.wolfSSL_CTX_new(ftab.wolfTLSv1_2_client_method());
+	if (!wolf_ctx) {
+		fprintf(stderr, "ERROR: failed to create WOLFSSL_CTX\n");
+		ftab.wolfSSL_Cleanup();
+	}
+
+	if (ctx->is_server) {
+		ftab.create_key_and_x509(wolf_ctx);
+		if (ctx->mutual) {
+			ftab.wolfSSL_CTX_set_verify(wolf_ctx,
+						    WOLFSSL_VERIFY_PEER |
+						    WOLFSSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+						    cert_verify_callback);
+		}
+	} else {
+		if (ctx->mutual) {
+			ftab.create_key_and_x509(wolf_ctx);
+		}
+		ftab.wolfSSL_CTX_set_verify(wolf_ctx, SSL_VERIFY_PEER,
+					    cert_verify_callback);
+	}
```

Comment:
Is it possible to unify the calling to set_verify? What the meaning of WOLFSSL_VERIFY_FAIL_IF_NO_PEER_CERT?